### DISCUSSION TITLE
Potential Miscalculation in Rebalancing Logic Leads to Liquidity Errors

### DISCUSSION THREAD

**Discussion by eshyun** [dodid/minitrade#9] [Q&A]
Hello,
I think Minitrade is an excellent project that aligns very well with my needs. I'm grateful for the hard work and dedication you've put into building such a useful tool.

I'm playing with Multi-Asset Example in the documentation.
The example itself which uses bucket.weight_explicitly() works flawlessly, however, bucket.weight_equally() raises error:
_RuntimeError: Not enough liquidity for <Order AAPL size=233.0, contingent=0>, has 9310, needs 9344, aborting_

From my observation, it seems that the rebalancing logic in minitrade might attempt to sell existing assets to free up capital for purchasing required assets. However, there might be a calculation issue in determining the exact amount of capital that can be reallocated. Specifically, the framework might be overestimating the available liquidity after sales, leading it to place orders that exceed actual available funds. This results in a “Not enough liquidity” error.

Would appreciate any guidance on this issue or insight into the rebalancing calculation approach to confirm if there is a workaround or planned fix. Thank you!

Source code
```python
from minitrade.backtest import Strategy
from minitrade.backtest import Backtest
from minitrade.datasource import QuoteSource

class MyStrategy(Strategy):
    lookback = 10

    def init(self):
        self.roc = self.I(self.data.ta.roc(self.lookback), name='ROC')     #1

    def next(self):
        self.alloc.assume_zero()                                #2
        roc = self.roc.df.iloc[-1]                              #3
        (self.alloc.bucket['equity']                            #4
            .append(roc.sort_values(ascending=False), roc > 0)  #5
            .trim(3)                                            #6
            .weight_equally()                                   #7
            .apply())                                           #8
        self.rebalance(cash_reserve=0.01)                       #9


yahoo = QuoteSource.get_source('Yahoo')
data = yahoo.daily_bar('AAPL,MSFT,NVDA,TSLA', start='2018-01-01', end='2019-01-01')

bt = Backtest(data, MyStrategy)
result = bt.run()
print(result)
```

Error stack
```
Traceback (most recent call last):
  File "/Users/apollo/projects/trading/portfolio/test/mttest.py", line 26, in <module>
    result = bt.run()
             ^^^^^^^^
  File "/Users/apollo/projects/trading/portfolio/.venv/lib/python3.11/site-packages/minitrade/backtest/core/backtesting.py", line 2029, in run
    broker.next()
  File "/Users/apollo/projects/trading/portfolio/.venv/lib/python3.11/site-packages/minitrade/backtest/core/backtesting.py", line 1551, in next
    self._process_orders()
  File "/Users/apollo/projects/trading/portfolio/.venv/lib/python3.11/site-packages/minitrade/backtest/core/backtesting.py", line 1693, in _process_orders
    raise RuntimeError(
RuntimeError: Not enough liquidity for <Order AAPL size=233.0, contingent=0>, has 9310, needs 9344, aborting
```

**Comment by dodid** [comment]
This issue is not necessarily an error but a result of how we calculate asset allocation: we use the previous day's end-of-day (EOD) prices to determine order quantities, then execute these orders at the next day’s market open. Since the quantities are based on EOD prices, but executed using opening prices, a significant gap up could mean insufficient cash to buy the planned quantities. The [rebalance()](https://dodid.github.io/minitrade/api_backtest/#minitrade.backtest.core.backtesting.Strategy.rebalance) function includes a cash_reserve parameter to set a cash buffer, helping to mitigate this risk.
